Gallery view gestures attempt №2

K1rakishou opened the pull request:
Previous one was actually fine but one little problem needed to be resolved, which wasn't. This attempt fixes that problem and another one (one image could be saved multiple times via the swipe down gesture).

This should have been done way earlier because tap/double tap to close works really bad due to the zooming and other shit.

Adamantcheese committed:
Adds swipe up to close and swipe down to save image options for the paged image viewer

K1rakishou committed:
Another attempt to add swipe up/down gestures to the image gallery

- This time the zoomed image inertia problem has been fixed! Should be little bit better than it was before.
- Also a check to prevent multiple saving of the same image has been added.

K1rakishou committed:
Fix comments

K1rakishou committed:
Make swipe up/down detection display agnostic and fix a bug when some images couldn't be closed by swiping due to incorrect default scale value

K1rakishou committed:
Increase a little bit the zoomIn difference threshold so when we zoom out back we don't have to be super precise

K1rakishou commented:
~~This PR should be merged after https://github.com/Adamantcheese/Kuroba/pull/449 this one.
After they are both merged this change (https://github.com/K1rakishou/Kuroba/commit/ffb35014268954ae5198d8a202229044f30f5981#diff-24702c3f2446edd364337cb6657b47e5R551) should be added as well because otherwise there won't be any gestures for video when webm streaming is turned on. Right now it's impossible.~~

K1rakishou committed:
A more proper handling of gallery gestures

So now we can swipe the image top/bottom even if we are currently zoomed in into the image BUT only if we are touching the opposite side of the image bound

K1rakishou committed:
Fix swipe gestures and videos/gifs

K1rakishou commented:
@Adamantcheese 

By the way, do you think we should allow swipe-to-bottom gesture (to download a file) when zoomed in into an image? Right now both gestures work. If you are touching either top or bottom of an image while zoomed in and do the swipe top/bottom then one of the gestures will be triggered. But I think that in such case both gestures should close the image. In other words there shouldn't be the "swipe to download" gesture when you are zoomed in. Because it's kinda strange.

Adamantcheese commented:
No, not while zoomed in. That caused issues last time.

K1rakishou commented:
@Adamantcheese You mean both swipes, top and bottom, should close the image or that we shouldn't allow gestures when zoomed in at all? Because if it's the latter then there shouldn't be any problems with them now. (IIRC, the last time there just wasn't the inertial scrolling when zoomed in because swipe gesture detector was always returning true when zoomed in. In this PR that behavior is fixed.)

Adamantcheese commented:
Better to have no gestures while zoomed in.

K1rakishou commented:
But that thing already has gestures to swipe left/right to open previous/next image. I think you should just try it out for yourself  because I think it works pretty good. 

Do the following: open an image and slightly zoom in (like 1-5%) then swipe to close. Then disable that thing and do it again. You will see that it's kinda frustrating that you can't swipe to close when an image is just slightly zoomed in.

Adamantcheese commented:
If I'm zoomed in and I swipe up or down, I expect to be flung to the top/bottom of the image. You can double tap to zoom back out and gestures would work again just fine.

K1rakishou commented:
>I expect to be flung to the top/bottom of the image

That's what will happen. The gesture will only be triggered if you are **already** touching top/bottom of the image with the viewport and do the swipe.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java:
```diff
@@ -103,45 +105,66 @@
-            return true;
-        }
-    });
-    private GestureDetector gifDoubleTapDetector = new GestureDetector(context, new SimpleOnGestureListener() {
```
Uh, why is this one being removed? You're removing the pause functionality for gifs.

K1rakishou replied:
Oh, damn. And the webms too. I will revert it back.

K1rakishou replied:
Done.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java:
```diff
@@ -103,45 +105,66 @@
     private SimpleExoPlayer exoPlayer;
-
     private boolean backgroundToggle;
+    private boolean imageAlreadySaved = false;
```
If you're going to prevent duplicate saves using this, you should probably also set the toolbar item to be greyed out as well.

K1rakishou replied:
Then I need FileCacheV2 to get merged to use it's toolbarMenuItem.enable() method. I don't want to copy it here from that PR. So this one will have to wait.

K1rakishou replied:
Wait a minute. This already works. The only more thing that was need to be implemented is to set this flag to true when we download an image from the ImageViewController to avoid double saves.

Adamantcheese reviewed (commented):
Noticed some other things.

K1rakishou commented:
@Adamantcheese So what do we do with the swipe-to-save image while zoomed in? Personaly, I think that both gestures should close the image while you are zoomed in.

Adamantcheese commented:
Either both gestures unzoom the imafe or both gestures close the image. That one might be best to test out beforehand.

K1rakishou commented:
@Adamantcheese since there is already double-tap to unzoom the image I will then do both gestures to close the image.

>That one might be best to test out beforehand.

Well, I do usually test my PRs. Doesn't matter whether they are merged or not.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java:
```diff
@@ -103,45 +105,66 @@
+
+    private static final float FLING_DIFF_Y_THRESHOLD = dp(60f);
+    private static final float FLING_VELOCITY_Y_THRESHOLD = dp(200f);
+    private static final float FLING_DIST_X_THRESHOLD = dp(100f);
```
I think the horizontal devitaion should be like 60 or 70dp. 100 is kinda too much. It allows the gesture to deviate more to left/right. And it shouldn't.

K1rakishou replied:
Done.

K1rakishou committed:
More improvements for gallery gestures

- Play/pause gif by single tapping was reverted back. Now you can also do them same with videos.
- Swipe-to-close gesture was added to thumbnails.
- Gesture thresholds have been tweaked a little bit to make it more harder to trigger the swipe top/bottom gestures (it was conflicting with double-tap-to-zoom-in gesture in SubsamplingScaleImageView).
- New behavior for swipe-to-save gesture: if an image is already saved then the next swipe-to-save will trigger swipe-to-close instead (which of course will be reset if you reopen the same image again because this flag is not persisted yet).
- When zoomed in on an image swipe-to-save will trigger swipe-to-close instead.
- Update the "imageAlreadySaved" flag when saving an image from ImageViewController.

K1rakishou commented:
~~**Reminder for myself**: Do not forget to merge this - https://github.com/K1rakishou/Kuroba/commit/ffb35014268954ae5198d8a202229044f30f5981#diff-24702c3f2446edd364337cb6657b47e5R551~~

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageViewGestureDetector.kt:
```diff
@@ -0,0 +1,155 @@
+    override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
+        val gifImageView = callbacks.findGifImageView()
+        if (gifImageView != null) {
+            // If a GifImageView is visible the toggle it's play state
```
then

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageViewGestureDetector.kt:
```diff
@@ -0,0 +1,155 @@
+                    bigImageView.isViewportTouchingImageBottom &&
+                    bigImageView.isViewportTouchingImageTop
+            ) {
+                // We are either not zoomed it or just slightly zoomed in (i.e. the view
```
zoomed in

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageViewGestureDetector.kt:
```diff
@@ -0,0 +1,155 @@
+            callbacks.onSwipeToCloseImage()
+        } else {
+            callbacks.onSwipeToSaveImage()
+            // TODO: disable the "save image" toolbar menu item once FileCacheV2
```
This is already done.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageViewGestureDetector.kt:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private fun onSwipedUp(bigImageView: CustomScaleImageView?): Boolean {
+        // If either other than big image view is visible (thumbnail, gif or video) OR
```
If either any view, other than big image view, is visible

K1rakishou committed:
Merge branch 'cheese-multi-feature' into gallery-view-gestures

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/presenter/ImageViewerPresenter.java
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java

K1rakishou committed:
Add gesture detector for exo player view with when webm streaming is on

K1rakishou committed:
Couple comment fixes

K1rakishou committed:
Merge branch 'cheese-multi-feature' into gallery-view-gestures

# Conflicts:
#	Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java

K1rakishou commented:
It was bugging me the whole week, so I decided to change swipe-to-save a little bit before it got merged.

K1rakishou committed:
A little improvement for swipe-to-save gesture

The previous swipe-to-save detection was not ideal. If you have a wide image (width > height) and you slightly zoom in you might expect that swipe-to-save would be replaced with swipe-to-close gesture but that wasn't true. That's because the viewport would still touch both the top and the bottom sides of the image. So to actually trigger swipe-to-close instead of swipe-to-save you would have to zoom further in until you can start panning the viewport vertically. While testing this thing it turned out to be a little bit irritating. So instead, it is now checking whether all sides of the viewport are touching all sides of an image before triggering swipe-to-save instead of swipe-to-close. So, basically, it is now only allowed when you are fully zoomed out (the default image state when you open it).

Adamantcheese commented:
I don't know why I didn't mention this before, but last time people kinda complained super hard that there was no option to turn off gestures. Either because they don't know how to use their phone correctly or whatever, it might be a good idea to add in that option to experimental settings for the time being.

Adamantcheese commented:
If you quickly toss in that setting, I think that it'd be fine to merge this in with this release actually.

K1rakishou commented:
Oh, yeah, we probably should. Alright, gonna do it right away. And with gestures being turned off there should only be the option to close media by clicking it right?

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/ThumbnailImageView.kt:
```diff
@@ -0,0 +1,11 @@
+import android.util.AttributeSet
+import android.widget.ImageView
+
+class ThumbnailImageView(
```
If you're overriding a class, it better do something other than just override it. You have a habit of doing that with other things as well, like when I tore apart your CancellableToast class and removed the dumb duplicated enum you put in. Just use the base class; none of the other views in here extend ImageView anyways.

K1rakishou replied:
No, I need a separate class that extends from ImageView so I can tell it apart from the other ImageView because we already have an ImageView that is being used for video controlls.

K1rakishou replied:
In other words findThumbnailImageView() won't work because it will return two ImageViews then and I only need to find one (the thumbnail)

K1rakishou replied:
https://github.com/Adamantcheese/Kuroba/pull/491/files/d21bec5be8b2933dd27540d8478bc3ac830f3f73#diff-24702c3f2446edd364337cb6657b47e5R150

Adamantcheese replied:
You can check to see if it's the playView ImageView instance in your search method and continue searching if it is. You do not need a separate class.

K1rakishou replied:
That sounds kinda unnatural. I think that using a separate class is kinda more elegant way to deal with that.

Adamantcheese replied:
I'll rip it out myself then. I don't want ANOTHER thumbnail view class when we already have three of them.

K1rakishou replied:
Alright, then how about I create a new class for video overlay instead? Like VideoOverlayImageView?

Adamantcheese replied:
Consider looking at PostImageThumbnailView before you start duplicating stuff again.

K1rakishou replied:
I should have probably just named it differently. But it's hard to come up with new names. So what about extracting the view overlay into it's own class?

K1rakishou replied:
PostImageThumbnailView contains some custom logic. I don't think we want getting an image url by longtapping it in gallery.

Adamantcheese replied:
Then maybe make it more generic? Add a flag to the constructor that would prevent that?

K1rakishou replied:
I just wanted to add "let's not add new flags to it" but I was too late, damn. See, it already has PostXXX in it's name, so it's not suitable for the gallery.

K1rakishou replied:
And we already use ThumbnailView for pins.

K1rakishou replied:
And they all have custom logic that is only suitable for the place they are being used in right now. Trying to reuse them in different places is just not natural. The best thing to do here is to just create a new class for gallery thumbnails. No flags needed, no additional checks that will break some time in the future. If you ever want to add something new to the gallery thumbnail it's super simple because it lives in it's own class, no need to extract it into a separate one or adding even more checks.

Adamantcheese replied:
Then I would recommend moving the playIcon stuff into that thumbnail view, so that it has some use other than just being an empty wrapper class.

K1rakishou replied:
You mean playIcon in PostImageThumbnailView? Then FixedRatioThumbnailView, which is used for pins, will also have playIcon and we probably don't want to see that icon on top of the pins' images.

Adamantcheese replied:
No, I mean move the playIcon stuff from MultiImageView into your new class. The end result should be similar to PostImageThumbnailView, but without the custom functionality, which is why I recommended looking at it in the first place.

K1rakishou replied:
I think we shouldn't use ThumbnailImage in gallery whatsoever. ThumbnailImage is used alongside the ImageLoader (it uses Volley's callbacks) and we don't use Volley in gallery at all. That would just be kinda strange. And, what's more important, we want to get rid of volley because there is not point in having two separate http libraries.

Adamantcheese replied:
Good sweet lord how am I not explaining this correctly. I'll just do it myself.

K1rakishou replied:
Oh, I get it now. You meant to create a new class with just the playIcon. Kinda like what I meant when suggesting extracting the video overlay into it's own class. Yeah, that sounds reasonable.

K1rakishou replied:
But we won't be able to use it with PostImageThumbnailView anyway since it already inherits from ThumbnailView which has tons of custom logic. So we will still have two 
 separate classes with playIcon. If that's what you meant.

K1rakishou replied:
>I'll just do it myself.

Please, just make sure that all findXXX() methods in MultiImageView can only find one class each because everything will break otherwise.

Adamantcheese replied:
Yes, that's what I meant. And I know that they should only find what they're looking for, which is why I suggested the nice and simple check instead of a refactor.

K1rakishou replied:
> the nice and simple check

Uhhh.

K1rakishou replied:
It's not nice and we will definitely forget about it in the future. Flags and checks are bad. And shared mutable state too.

K1rakishou committed:
Add an experimental setting to turn gestures on/off (on by default)

K1rakishou commented:
Alright. The setting is in.

K1rakishou committed:
Merge branch 'cheese-multi-feature' into gallery-view-gestures

# Conflicts:
#	Kuroba/app/src/main/res/values/strings.xml

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java:
```diff
@@ -261,11 +303,6 @@ public void setVolume(boolean muted) {
         }
     }
 
-    @Override
```
It's weird, but this should probably stay.

K1rakishou replied:
Well, if you think it should be there to allow closing an image by clicking outside of it (like the view's transparent background area) then it works even without that callback.

Adamantcheese replied:
At the very least, tie it to the setting you added for gestures.

K1rakishou replied:
But it works even with gestures being turned off.

Adamantcheese replied:
Hrm, alright.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/view/MultiImageView.java:
```diff
@@ -128,8 +123,13 @@
+    private boolean imageAlreadySaved = false;
+
+    private GestureDetector gestureDetector = new GestureDetector(
+            getContext(),
```
You probably should put this in the constructor, I'm not sure if calling getContext() before the constructor would cause a weird bug or not, because the order of construction per java specification is initialize class variables -> call the constructor.

K1rakishou replied:
Done.

K1rakishou committed:
Code-review fixes
